Checkout from GitHub only once per branch, stash and then just unstash

ronald-cron-arm opened the pull request:
Fixes #197 and reduce significantly the number of requests to GitHub.
- OpenCI
  - TF-PSA-Crypto
    - [x] [PR-head](https://mbedtls.trustedfirmware.org/job/mbed-tls-tf-psa-crypto-restricted-ci-testing/view/change-requests/job/PR-3-head/22/)
  - Mbed-TLS
    - development
      - [x] [PR-head](https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-ci-testing/view/change-requests/job/PR-906-head/36/)
      - [x] [PR-merge](https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-ci-testing/view/change-requests/job/PR-906-merge/56/)
    - 3.6
      - [x] [PR-head](https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-ci-testing/view/change-requests/job/PR-1233-head/11/)
      - [x] [PR-merge](https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-ci-testing/view/change-requests/job/PR-1233-merge/17)
    - release
      - [x] [development + 3.6](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/368/)

- Internal CI
  - TF-PSA-Crypto
    - [x] [PR-head](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-tf-psa-crypto-restricted-ci-testing/view/change-requests/job/PR-3-head/14/)
  - Mbed-TLS
    - development
      - [x] [PR-head](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-ci-testing/view/change-requests/job/PR-906-head/135/)
      - [x] [PR-merge](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-ci-testing/view/change-requests/job/PR-906-merge/162/)
    - 3.6
      - [x] [PR-head](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-ci-testing/view/change-requests/job/PR-1233-head/14/)
      - [x] [PR-merge](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-ci-testing/view/change-requests/job/PR-1233-merge/13/)
    - release
      - [x] [development + 3.6](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/805/)

bensze01 commented on vars/common.groovy:
```diff
@@ -333,8 +334,15 @@ List<BranchInfo> get_branch_information(Collection<String> tls_branches, Collect
+
+    def results = [:]
+    results.putAll(all_results)
+    results.putAll(platform_results)
```
We can simplify this slightly
```diff
diff --git a/vars/common.groovy b/vars/common.groovy
index 268bfb6..2e5682c 100644
--- a/vars/common.groovy
+++ b/vars/common.groovy
@@ -340,14 +340,10 @@ List<BranchInfo> get_branch_information(Collection<String> tls_branches, Collect
     platform_jobs.failFast = true
     def platform_results = (Map<String, Map<String, String>>) parallel(platform_jobs)

-    def results = [:]
-    results.putAll(all_results)
-    results.putAll(platform_results)
-
     infos.each { BranchInfo info ->
-        info.all_sh_components = results[info.job_prefix + 'all-platforms']
+        info.all_sh_components = all_results[info.job_prefix + 'all-platforms']
         linux_platforms.reverseEach { platform ->
-            info.all_sh_components << results[info.job_prefix + platform]
+            info.all_sh_components << platform_results[info.job_prefix + platform]
         }

         if (env.JOB_TYPE == 'PR') {
```

ronald-cron-arm replied:
Indeed, thanks.

bensze01 reviewed (dismissed):
One small nit, otherwise looks good to me.

gilles-peskine-arm commented on vars/checkout_repo.groovy:
```diff
@@ -150,15 +150,28 @@ git submodule foreach --recursive git config url.git@github.com:.insteadOf https
+            }
+
+            def stashName = "${info.job_prefix}stash"
+            stash name: stashName, includes: '**/*', useDefaultExcludes: false
```
If `checkout_repo` is invoked with the same `info.job_prefix` value in parallel, there is a race condition: whoever reaches the `stash` call first will error out because the stash name already exists.

This doesn't seem to be a problem given the way we currently use `checkout_repo`, but it looks fragile to me. For example, maybe it's a problem only if we run certain subsets of the jobs? Please at least document it, though I'd feel more comfortable with a lock.

ronald-cron-arm replied:
By construction of `info.job_prefix` it is not possible to have different `info` objects with the same `job_prefix`. Thus there is no need for a lock as the code stands. But if you still prefer I can add a lock. Do we have a example of that?

gilles-peskine-arm replied:
> By construction of info.job_prefix it is not possible to have different info objects with the same job_prefix

This is not at all obvious to me, and I can't see it documented anywhere. If it's an expected invariant, I'd like it to be documented. Even if it's the case now, given the lack of documentation, it may not be the case after we add some feature or other.

> But if you still prefer I can add a lock. Do we have a example of that?

Going by example in `gen_jobs.groovy`, I guess something like
```
lock("checkout-repo/${info.job_prefix}") { … }
```

ronald-cron-arm replied:
I've tried to improve the code and the documentation rather than adding a lock.

gilles-peskine-arm replied:
Thanks for improving the documentation and cleaning up the code! But it's still not clear to me why `checkout_repo` would never be called concurrently on the same repo+branch. Actually, that does happen, doesn't it? What makes it work is that the _first_ call for a given repo+branch is always done during some setup phase that is unique and must complete before the repo+branch is used again. That property seems pretty fragile (e.g. we might break it if we add a Python setup phase that's independent of the `all.sh` gathering phase), and it's undocumented.

ronald-cron-arm replied:
That's a good point, thanks. I will improve that.

gilles-peskine-arm commented on vars/checkout_repo.groovy:
```diff
@@ -150,15 +150,28 @@ git submodule foreach --recursive git config url.git@github.com:.insteadOf https
+            stash name: stashName, includes: '**/*', useDefaultExcludes: false
+            info.stash = stashName
+        } else {
+            unstash info.stash
```
The [Jenkins documentation](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#stash-stash-some-files-to-be-used-later-in-the-build) says:

> Note that the stash and unstash steps are designed for use with small files. For large data transfers, use the External Workspace Manager plugin, or use an external repository manager such as Nexus or Artifactory. This is because stashed files are archived in a compressed TAR, and with large files this demands considerable resources on the controller, particularly CPU time. There's not a hard stash size limit, but between 5-100 MB you should probably consider alternatives.

An mbedtls-3.6 source tree with the framework is about 27MB. Should we fear performance problems?

Note that we already use stashes for outcomes, for which I've also wondered this and never gotten an answer. Those are mostly about 8MB apiece, with a few larger ones. So if the size is a problem, it's just making an existing problem worse.

ronald-cron-arm replied:
It seems you are right we are likely to overload the controller. @bensze01 do you confirm? It seems an alternative is to use an "Artifact Manager (e.g., S3, Artifactory)". @bensze01 do you know if we can use that in the internal and Open CI?

ronald-cron-arm replied:
After discussions with Bence we think it is worth giving it a try. The controller does not seem a bottleneck currently and the size of a stash (~50M) is not that big. If it does not scale, it is easy to revert and we can then try to use an "S3 bucket" to handle the unstash traffic in place of the controller.

gilles-peskine-arm commented on vars/common.groovy:
```diff
@@ -246,7 +247,7 @@ List<BranchInfo> get_branch_information(Collection<String> tls_branches, Collect
             }
 
-            jobs << gen_jobs.job(info.job_prefix + 'all-platforms') {
+            all_jobs << gen_jobs.job(info.job_prefix + 'all-platforms') {
```
I don't understand the variable names `all_jobs` and `platform_jobs` in this function. I think how `platform_jobs` has one entry per platform, so that sort of makes sense. But `all_jobs` is neither “all the jobs” in some sense, nor “jobs related to `all.sh`”, and I'm confused.

ronald-cron-arm replied:
I've tried to improve the variable names.

gilles-peskine-arm reviewed (changes requested):
Looks good to me functionally (pending CI results for the latest commit), but I have a couple of concerns about performance and concurrency.

gilles-peskine-arm reviewed (commented):
LGTM on code reading at 479ce7e3ac28464119db6839600a53c755f3c0eb except that I'd like documentation or code to address https://github.com/Mbed-TLS/mbedtls-test/pull/201#discussion_r2063207288.

bensze01 commented on vars/gen_jobs.groovy:
```diff
@@ -458,10 +458,28 @@ def gen_abi_api_checking_job(BranchInfo info, String platform) {
+                esac
+
+                git fetch --depth 1 origin ${CHANGE_TARGET}
+            '''
```
This can be simplified down to a single-line:
```suggestion
sh "GIT_SSH_COMMAND='ssh -o StrictHostKeyChecking=accept‐new' git fetch --depth 1 origin ${CHANGE_TARGET}"
```

ronald-cron-arm replied:
This is indeed much shorter but our CI does not support the "accept-new" option. Otherwise, in terms of security it is not as good as with this approach we trust blindly the first connection. Actually I'd say there is no authentication at all but I may miss something.

bensze01 commented on vars/checkout_repo.groovy:
```diff
@@ -150,23 +150,32 @@ git submodule foreach --recursive git config url.git@github.com:.insteadOf https
-                    break
-                default:
-                    error("Invalid repo: ${info.repo}")
+            lock(resource: "checkout_repo-${stashName}", inversePrecedence: true) {
```
Use local per-branch locks instead of global ones
```suggestion
            synchronized (info) {
```

ronald-cron-arm replied:
Doing some researches it seems that this does not work in the context of Jenkins pipelines as Jenkins Pipelines are not multi-threaded in the JVM sense.

bensze01 replied:
Are you sure?
Jenkins readily offers thread dumps in the java sense for parallel stages.
I was under the impression that Jenkins pipelines basically have a Global Interpreter Lock like Python, which they release for long-running steps like `checkout` or `sh`.

So `synchronized` should work just fine.

bensze01 reviewed (changes requested):
I've requested two changes simplifying the code and reducing resource contention.
Otherwise looks good to me.

gilles-peskine-arm commented on vars/gen_jobs.groovy:
```diff
@@ -427,50 +427,76 @@ def gen_windows_testing_job(BranchInfo info, String toolchain, String label_pref
+                            domain=$(printf "%s" "$origin_url" | sed -n 's#ssh://[^@]*@\\([^/:]*\\).*#\\1#p')
+                            echo "origin remote domain: $domain"
+                            mkdir -p ~/.ssh
+                            ssh-keyscan "$domain" >> ~/.ssh/known_hosts
```
This makes me uncomfortable. We're potentially enabling man-in-the-middle attacks. Why do we need this? It doesn't seem related to the rest of the PR.

bensze01 replied:
This is what the OpenCI is configured to do at the moment when we do checkouts.
But this change is applied on a per-node basis, and we don't call `checkout` on all nodes anymore, leading to intermittent failures (since there is no interactive user to type `yes` when ssh sees a new server)
Our alternative solution is to ask and wait for Linaro to hard-code Github's public key on our AMI machine images.

gilles-peskine-arm replied:
Ok, thanks for the information, but why wasn't it a problem before? It's not that we were updating `known_hosts` only on the master before and now we do it on all the nodes. If the master has reliable information and the nodes don't, then we should copy the information from the master rather than obtain it in an unreliable way.

bensze01 replied:
`checkout` uses `ssh` checkouts directly on the nodes.
We had issues before when Github updated their `ssh` public keys (due to a breach), and after that Linaro configured Jenkins to `accept new keys for unknown hosts`

This means that when calling `checkout` jenkins is configured to accept and add the public key of any new/unkown host to `known_hosts` on the node the checkout is running on.

But this file gets reset every time the node spins down due to inactivity.

Previously we always had at least one `checkout` on the same node that the `˙git fetch` was called on, so the `known_hosts` was pre-populated. But this PR gets rid of this, and only uses `unstash` before `fetch` - so if we get scheduled on a "fresh" node that had no `checkout` steps executed on it since it spun up, we get an error.

bensze01 replied:
I've filed the issue:
https://linaro.atlassian.net/browse/TFC-747

bensze01 replied:
@ronald-cron-arm I would actually prefer that we hardcode `github.com` into the `ssh-keyscan` workaround instead of trying to parse the origin url, since we don't use any other hosts, and we expect that this will be short-term workaround until Linaro fix this on their side.

ronald-cron-arm replied:
I've removed that change.

gilles-peskine-arm reviewed (changes requested):
LGTM on code reading except I don't understand the SSH thing.

ronald-cron-arm commented:
I've rebased following the merge of #203 and removed the commit about the GitHub URL. The potential issue with `gen_abi_api_checking_job` and parametrized jobs will be addressed through  https://linaro.atlassian.net/browse/TFC-747.

gilles-peskine-arm reviewed (approved):
LGTM at b4b957ede246ee6f2fea0749709a84a2985a48fa on code reading.

bensze01 reviewed (approved):
Looks good to me, pending the CI results.

ronald-cron-arm commented:
All tests have run successfully. I am merging it then.
